ui/raidboss/popup-text.ts

```diff
@@ -788,6 +793,18 @@ export class PopupText {
     this.data = this.getDataObject();
     this.StopTimers();
     this.triggerSuppress = {};
+
+    for (const init of this.dataInitializers) {
+      const data = init();
+      if (data) {
```

Comment:
I either don't understand what this is guarding or why we're guarding it. I understand this is for a scenario where we somehow have a malformed function, but it feels like this shouldn't be possible aside from a user file override, in which case this should already bubble up an issue? Maybe there's something else here that I'm missing?

Reply:
This is specifically for user file overrides. Because those are javascript-only, we can't rely on the typing at all, so if they return undefined or something we don't want to cause that to error out. Maybe it should be checking `if(isObject(data))` instead, but I think this check is required regardless.

Reply:
So, I guess my mental model is that a user override wouldn't really need/want this? Or care about it? And, we should get a bigger error (in general) that something bad happened? In either sense, I guess I'm proposing dropping this or making it very explicit that something failed within `initData`.

Is there a use-case where a user might want an extremely complex initData function that actually has data initialized? For most cases, I was just assuming this was a "clear having to check undefined 100 times in a TypeScript file" which doesn't apply for JavaScript user files

Reply:
My thought process here is as follows:

1. There's no good way to distinguish between a user trigger set and a built-in trigger set.
2. `initData` is available on both
3. Therefore, a user may use `initData` in a custom trigger set
4. If a user uses `initData` and doesn't return a proper object, we don't want to completely break raidboss for them

Reply:
Right, but then the user isn't getting what they expect, and there's a log about data initializers which might not explicitly tell them that the thing they did doesn't work? I'm thinking in the sense of a user gets another user's user files and has no idea what they're doing, then they drop those user files into their user directory and it "doesn't work" it's a lot greyer of an area to support than an explicit log that says, "ERROR IN USER FILE OVERRIDE INIT DATA", and doesn't have nearly the same impact as crashing entirely from "I need to back these out immediately" or "I added a file and now cactbot doesn't work. HELP"

Reply:
I guess I don't understand the question/concern here. If the `if(data)` check isn't present, then a bad user-supplied initData function will clobber the raidboss data object and cause every other trigger to break.

So either we:

- Don't allow user-supplied initData functions at all, which would require some sort of additional checking to not push them to the array, and goes against the data model
- Allow user-supplied initData functions, but keep users from breaking their raidboss entirely, and instead warn them and ignore the bad initData function
- Allow user-supplied initData functions, and let users completely break their raidboss

Right now, the second above is what is in this PR.

Reply:
I do think checking `if (typeof data === 'object')` instead of `if (data)` would be both clearer to the reader here and would catch more errors.

Reply:
After some more discussion/thought, I think one missing piece from the error handling here is the filename (which I guess would need to be tracked per init function, sorry).  Having something closer to ``console.log(`Error in file: ${filename}: these triggers may not work; initData function returned invalid object: ${init.toString()}`);``

But, I think it would be helpful to less technical users who aren't trigger authors (possibly copying and pasting triggers from friends) to know which file is busted and what the result of that might be.

Reply:
Sorry, to follow up on this (work got busy): I am not suggesting the first point at all; I don't think that's viable. I think latter two points are both okay, and both have their benefits. I think that if you're going to go with the second approach (as is currently in the PR), you need to be very explicit with the fact that there is an error. I would be surprised if more than 1% of users are reading the cactbot logs, so you need an approach that makes it either fairly obvious to the user that they did something wrong (the third point, hopefully), or something fairly obvious to whatever technical support (Discord, GitHub, etc.) that there were issues. I think what @quisquous has above works well; I think as long as the words `Error` and `initData` come across (filename super helpful too!) then it shouldn't be impossible to navigate from there.

Reply:
See latest commits, please. Sorry for the rebase/force-push, I wanted to pick up the webpack config changes so I could actually run this in dev server and test properly.

Specifically this commit should address the issues?

https://github.com/quisquous/cactbot/pull/3081/commits/42ae20988ad0450d0c4e56faf369406d5f77bd1e